### ISSUE TITLE
Updating VS 2019 from 16.3.x to 16.14.0 uninstall existing .NET Core SDK 3.0.100

### ISSUE THREAD

**Issue by eriawan** [dotnet/sdk#4015]
I have 2 different machines that has VS 2019 16.3.10 and VS 2019 16.3.7 on Windows 10 1909. 
Both of them have .NET Core SDK v3.0.100 installed.

After I update to VS 2019 16.4.0, the .NET Core SDK v3.0.100 got uninstalled and .NET Core 3.1.100 is installed. I don't mind having .NET Core 3.1.100 SDK installed. 

My problem is why .NET Core 3.0.100 is uninstalled on every time I update VS 2019 from v16.3.x to v16.4.0?

**Comment by johnbeisner** [comment]
@eriawan 
.NET Core 3.1.100 supports 3.0, 2.2, & 2.1
We have designed .NET Core with Visual Studio to only have a single SDK toolset - typically the most current released SDK. However, we are aware of an installation BUG that removes the 3.0 Runtimes in on this upgrade gesture - is this your concern?

**Comment by eriawan** [comment]
@johnbeisner 
Yes, that's my concern. Is there a planned schedule to fix this?

**Comment by johnbeisner** [comment]
@eriawan
We have fixed this for 16.5 and future releases; 16.3 and 16.4 will still have this update gesture problem. The workaround is to modify your VS installation and add back the desired Runtimes from the 'Individual component' list.

**Comment by NickCraver** [comment]
My team has hit this with every project as well, and we've had to alter every `global.json` in every repo for everyone that upgraded. It's not an awesome experience.

We have/had a `global.json` like so (in dozens of repos) like so:
```json
{
  "sdk": {
    "version": "3.0.100"
  }
}
```
...to which we had to add `"rollForward": "latestMajor"`. While it's great there's a mechanism for this, the experience on upgrading is really crappy for developers. The VS installer leaves the `C:\Program Files\dotnet\sdk\3.0.100` directory in place, so **it's a broken SDK**. e.g. this is what you got at the console after the VS upgrade:
```
λ dotnet --version
Found .NET Core SDK, but did not find dotnet.dll at [C:\Program Files\dotnet\sdk\3.0.100\dotnet.dll]
```
...that's a very confusing experience (and I'm hoping a bug - but I've hit this several times in upgrades now). Half of this is because it was removed, but it wasn't totally removed and since SDK version detection works on "is the folder there?", Visual Studio breaks pretty bad here and I've gotten a lot of requests for help from devs across the company on this. Here's their experience after the upgrade to 16.4:

![Error Dialog](https://user-images.githubusercontent.com/454813/70862862-c8ad6a00-1f0f-11ea-842d-d25a998c288c.png)

After they close that...

![Visual Studio](https://user-images.githubusercontent.com/454813/70862856-bcc1a800-1f0f-11ea-8bac-4aaabe3c659d.png)

I was a fan of the VS installer removing all of the *preview* releases when this was pitched. Huge fan. I racked up many gigs on SDKs. But in practice, this has caused more pain than it removed. I ask this: **Can we please only uninstall/cleanup the same minor version?**

Without that scope, I have to either go re-install the SDK (which really defeats the purpose here), or go change the `global.json` in every repo. Neither of those are good for me or my team, and this cleanup feature becomes a burden rather than a help.

**Comment by NickCraver** [comment]
cc @KathleenDollard @danroth27 ^ This is a follow-up to some discussions we had about upgrade pain points the whole team has felt, documenting it here.

**Comment by onovotny** [comment]
More in this thread https://twitter.com/onovotny/status/1206224178921246720

IMO, the current behavior is good and desirable. What might be missing is a way in global.json to opt-in/out of using preview SDK's but without specifying a version number. That way, if any preview SDK is on the machine, for a particular project, it'll fallback to using the latest stable instead of latest preview, etc.

**Comment by NickCraver** [comment]
I think it's also worth bringing up here: people came to me on how to fix this because the `global.json` docs still don't have any fixes or suggestions around this. `rollForward` remains undocumented AFAIK.

For example, there exists a `allowPrerelease` like Oren mentions above, but you _really_ have to be involved here to know that exists.

Here are the docs: https://docs.microsoft.com/en-us/dotnet/core/tools/global-json
And here is the issue about none of the 3.0 features being in there: https://github.com/dotnet/docs/issues/14479

**Comment by MeikTranel** [comment]
I don't think the documentation to that really is enough. I am almost certain that the majority of .net developers do not know anything about global.json at all. Most .net people in my social circle still think of the solution file as the center point to migrating tooling (which has some merit, which is making this experience even worse). 
I've been helping with a bunch of library authors migrating their tooling to new csprojs etc. and the overwhelming response was "meh i will not touch this file again for the next gazillion years... the sln file says 2019 now tho -> approve, squash and merge". This is not healthy at all.

**Comment by mairaw** [comment]
Regarding the docs, I have a PR in progress that I’m planning to finish this week: https://github.com/dotnet/docs/pull/15382

It needs more examples and explanations of the behavior changes starting in 3.0.

**Comment by mariusGundersen** [comment]
I'm having problems after upgrading to VS 16.4.1, and I don't have a global.json file. I have hundreds of errors in my small project, but it seems like the one that causes most of them is 
```
Error	NETSDK1073	The FrameworkReference 'Microsoft.NETCore.App' was not recognized	My.Project	C:\Program Files\dotnet\sdk\3.1.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets	283	
```

I have csproj files with `<TargetFramework>netcoreapp3.0</TargetFramework>` and the don't load correctly anymore. Cleaning does not seem to help, and setting it to use netcoreapp3.1 does not help either. I'm going to try installing 3.0.101 and see if that helps.

**Comment by abelbraaksma** [comment]
This behavior proves very problematic in practice. In the past, SDK's were left alone when updating, but at some point, SDK's were uninstalled when updating through the VS Installer. While this is trivial to fix once you know it (just download the SDK's by hand), it leads to very unwelcome surprises, like SDK-dependent solutions (through `global.json` etc) not loadable anymore inside VS.

I suggest this be fixed somehow. I can imagine making it a choice in the uninstaller (just ask the user), or going back to the previous behavior by simply leaving these SDK's alone. 

Adding it as an option when you select "Update" would probably be the best way forward.

**Comment by sake402** [comment]
Just to note that I updated VS2019 to version 16.8 and viola, all my existing SDKs gone and replaced with 5.0. 
Not everyone will be willing to jump right to dotnet 5, at least not until we are sure it is stable enough for production.

**Comment by redwyre** [comment]
Please stop uninstalling sdks, I updated vs and build tools on a build machine and it broke all my builds!

If you want to uninstall old SDKs, ASK FIRST

**Comment by velvetdust** [comment]
> This just happened AGAIN, please stop uninstalling sdks! I updated vs and build tools on a build machine and it broke all my builds!
same thing here:
`Found .NET Core SDK, but did not find dotnet.dll at [C:\Program Files\dotnet\sdk\5.0.202\dotnet.dll]`

**Comment by redwyre** [comment]
AND AGAIN

Updating removed 5.0.1xx when installing 5.0.3xx, breaking all existing builds again

**Comment by mairaw** [comment]
Tagging @timheuer

**Comment by KathleenDollard** [comment]
Visual Studio uninstalls what it installs on an update. I understand that it is causing an inconvenience, but the alternative is to have Visual Studio not clean up after itself that would also be bad. 

If you install via the [SDK installer by downloading]( https://dotnet.microsoft.com/download/dotnet) Visual Studio will not uninstall. We use normal Windows installer ref counting and separately ref count VS and the CLI, so after VS updates itself the ref count for the CLI remains and will block the uninstall of the earlier version.
